File: src/master/Master.java
/**
 * Copyright (c) 2011 The Regents of The University of Michigan
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met: redistributions of source code must retain the above copyright
 * notice, this list of conditions and the following disclaimer;
 * redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution;
 * neither the name of the copyright holders nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * @author David Meisner (meisner@umich.edu)
 *
 */
package master;

import java.io.BufferedReader;
import java.io.DataInputStream;
import java.io.FileInputStream;
import java.io.InputStreamReader;
import java.rmi.RemoteException;
import java.util.Iterator;
import java.util.Vector;

import stat.StatisticsCollection;
import core.Experiment;

/**
 * A Master controls the overall state of the simulation and is in
 * charge of commanding slaves.
 *
 * @author David Meisner (meisner@umich.edu)
 */
public final class Master {

    /**
     * Information about simulation slave are saved in this vector. This
     * information is found in a .cfg file.
     */
    private Vector<SlaveInfo> slaves;

    /**
     * The starting time of an experiment.
     */
    private long startTime;

    /**
     * The master has its own experiment to run at the beginning of simulation.
     * This information is later passed onto slaves.
     */
    private Experiment masterExperiment;

    /**
     * The Experiments for each Slave in the simulation.
     */
    private Experiment[] slaveExperiments;

    /**
     * Constructs a new Master.
     */
    public Master() {
        this.slaves = new Vector<SlaveInfo>();
    }

    /**
     * Connects to all slaves the Master knows off.
     */
    public void connectAll() {
        Iterator<SlaveInfo> iter = this.slaves.iterator();
        while (iter.hasNext()) {
            iter.next().connect();
        }
    }

    /**
     * Checks that the Master can connect to its slaves.
     * Prints a simple output for each.
     */
    public void checkConnectivity() {
        Iterator<SlaveInfo> iter = this.slaves.iterator();
        while (iter.hasNext()) {
            SlaveInfo slaveInfo = iter.next();
            String returnedString = "";
            try {
                returnedString = slaveInfo.getInterface().sayHello();
            } catch (RemoteException e) {
                e.printStackTrace();
            }
            System.out.println("Master got output: " + returnedString + "#");
        }
    }

    /**
     * Adds SlaveInfo about a slave to the master.
     * @param serverName - The name to refer to the server by.
     * @param rmiBinding - A RMI binding id.
     */
    public void addSlave(final String serverName, final String rmiBinding) {
        SlaveInfo slaveInfo = new SlaveInfo(serverName, rmiBinding);
        this.slaves.add(slaveInfo);
    }

    /**
     * Gets info about all the slaves of the master.
     * @return An Iterator of SlaveInfo of all the slaves
     */
    public Iterator<SlaveInfo> getSlavesInfo() {
        return this.slaves.iterator();
    }

    /**
     * Gets the name of the Master node.
     * @return The name of the master.
     */
    public String getMasterName() {
        return this.slaves.firstElement().getServerName();
    }

    /**
     * Parses the configuration file specifying the master/slave setup.
     *
     * @param file - The configuration file
     * @return number of slaves in the configuration
     */
    public int parseConfigFile(final String file) {
        System.out.println("Parsing config file: " + file);
        final int partsPerLine = 4;
        final int instancesIndex = 3;

        try {
            FileInputStream fstream = new FileInputStream(file);
            DataInputStream in = new DataInputStream(fstream);
            BufferedReader br = new BufferedReader(new InputStreamReader(in));
            String strLine;

            while ((strLine = br.readLine()) != null) {
                // This line is a comment.
                if (strLine.startsWith("#")) {
                    continue;
                }
                // Valid lines have exactly PARTS_PER_LINE parts.
                String[] parts = strLine.split("\\s+");
                if (parts.length != partsPerLine) {
                    continue;
                }

                String host = parts[0];
                int instances = Integer.valueOf(parts[instancesIndex]);
                for (int i = 0; i < instances; i++) {
                    this.addSlave(host, "sim_" + i);
                }
            }

            in.close();
        } catch (Exception e) {
            System.err.println("Error: " + e.getMessage());
            e.printStackTrace();
        }

        return slaves.size();

    }

    /**
     * Locally run the experiment on the Master until it reaches steady state.
     * @param experiment - The experiment to run
     */
    public void runMasterExperiment(final Experiment experiment) {
        startTime = System.currentTimeMillis();
        masterExperiment = experiment;
        System.out.println("Master starting simulation");
        this.connectAll();
        System.out.println("Checking connectivity");
        this.checkConnectivity();
        System.out.println("Running to steady state on master");
        masterExperiment.runToSteadyState();
        System.out.println("Done running to steady state");
    }

    /**
     * This method runs the slave experiments.
     *
     * @param experiments - an array of experiments
     */
    public void runSlaveExperiment(final Experiment[] experiments) {
        //TODO Do we really need the experiments array?
        final long sleepTime = 10000;
        final double millisecondsPerSecond = 1000.0;
        slaveExperiments = experiments;

        System.out.println("Starting up slaves");
        this.startAllSlaves();
        boolean sleepHold = true;
        while (sleepHold) {
            try {
                Thread.sleep(sleepTime);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
            boolean done = true;

            Iterator<SlaveInfo> iter = this.slaves.iterator();
            StatisticsCollection combinedStats = null;
            System.out.println("Checking if combined we're done");
            while (iter.hasNext()) {
                SlaveInfo slave = iter.next();
                try {
                    StatisticsCollection stats = slave.getInterface()
                            .getExperimentStats();
                    stats.printAllStatInfo();
                    if (combinedStats == null) {
                        combinedStats = stats;
                    } else {
                        combinedStats = combinedStats.combine(stats);
                    }

                } catch (RemoteException e) {
                    e.printStackTrace();
                }
            }
            System.out.println("Combined info:");
            combinedStats.printAllStatInfo();
            System.out.println("combinedStats.allStatsConverged() =="
                    + combinedStats.allStatsConverged());
            if (!combinedStats.allStatsConverged()) {
                done = false;
            }

            if (done) {
                System.out.println("I think i'm done!");
                sleepHold = false;
            }
        }

        System.out.println("Done sleeping");
        this.combine();
        long endTime = System.currentTimeMillis();
        double execTime = (endTime - startTime) / millisecondsPerSecond;
        System.out.println("Combined Experiment time: " + execTime + " (s)");

        Iterator<SlaveInfo> iter = this.slaves.iterator();
        while (iter.hasNext()) {
            SlaveInfo slave = iter.next();
            try {
                slave.getInterface().stop();
            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }
    }

    /**
     * This method combines and outputs the final statistics.
     */
    public void combine() {
        System.out.println("***** Starting combine phase ****");
        StatisticsCollection totalStats = null;

        Iterator<SlaveInfo> iter = this.slaves.iterator();
        while (iter.hasNext()) {
            SlaveInfo slave = iter.next();
            try {
                StatisticsCollection stats = slave.getInterface()
                        .getExperimentStats();
                if (totalStats == null) {
                    totalStats = stats;
                } else {
                    totalStats = totalStats.combine(stats);
                }

            } catch (RemoteException e) {
                e.printStackTrace();
            }
        }

        System.out.println("========== Final Statistics ==========");
        totalStats.printConvergedOutputs();
        System.out.println("======================================");
    }

    /**
     * This method runs experiments on Slaves.
     */
    private void startAllSlaves() {

        int uniqueSeed = 2;
        Iterator<SlaveInfo> iter = this.slaves.iterator();
        int i = 0;
        while (iter.hasNext()) {
            SlaveInfo slave = iter.next();
            slaveExperiments[i].setSeed(uniqueSeed);
            System.out.println("Running experiment on " + slave.getServerName()
                    + "-" + slave.getRmiBinding());
            try {
                slave.getInterface().runExperiment(slaveExperiments[i]);
            } catch (RemoteException e) {
                System.out.println("Exception as string: " + e.toString());
                e.printStackTrace();

            }
            uniqueSeed++;
            i++;
        }
    }

}

File: master.py
#!/usr/bin/python

#
# Copyright (c) 2011 The Regents of The University of Michigan
# All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met: redistributions of source code must retain the above copyright
# notice, this list of conditions and the following disclaimer;
# redistributions in binary form must reproduce the above copyright
# notice, this list of conditions and the following disclaimer in the
# documentation and/or other materials provided with the distribution;
# neither the name of the copyright holders nor the names of its
# contributors may be used to endorse or promote products derived from
# this software without specific prior written permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
#
# Authors: Junjie Wu (wujj@umich.edu)
#

# This script defines and runs a distributed SQS experiment
# You shouldn't directly run this script from command line 
# Instead, you should use sqs.py to launch SQS, which will invoke this file
# This script uses JPype to access SQS Java library (Check README for JPype configuration)

# Most of the codes are JPype and SQS setups
# Due to the limition of JPype, everything has to happen in one file during one JVM session
# The only things you need to change to script an experiment are stat_config and createExperiment()

import os.path
import sys
import time

# JPype: import jar files
print "========== JPype Output Begin =========="
import jpype
jarpath = os.path.abspath('.')
jpype.startJVM(jpype.getDefaultJVMPath(), "-Djava.ext.dirs=%s" % jarpath)
print "========== JPype Output End ============"

# JPype: import java packages
java = jpype.JPackage('java')
sqs_core = jpype.JPackage('core')
sqs_datacenter = jpype.JPackage('datacenter')
sqs_generator = jpype.JPackage('generator')
sqs_math = jpype.JPackage('math')
sqs_stat = jpype.JPackage('stat')
sqs_master = jpype.JPackage('master')
sqs_StatName = jpype.JClass('core.Constants$StatName')
sqs_SocketPowerPolicy = jpype.JClass('datacenter.Socket$SocketPowerPolicy')
sqs_CorePowerPolicy = jpype.JClass('datacenter.Core$CorePowerPolicy')

# ========== Don't change code above unless you need to import new packages ==========

# ========== Change stat_config and createExperiment() to define you own experiment ==========

# global statistic requirements for convergence
# SQS will run until the aggregated result meets the statistical requirements
# stat_config is a list of target statistics, each element is a 5-tuple
# (stat name, mean precision requirement, quantile setting, quantile precision requirement, warmup samples)
# stat name is the output you are interested in. Supported statname is defined in src/core/Constants.StatName
# precision requirements are denoted by alpha value
# warmup samples are the number of samples discarded before collecting statistics
stat_config = [ (sqs_StatName.TOTAL_CAPPING, 0.05, 0.95, 0.05, 500),
				(sqs_StatName.SOJOURN_TIME, 0.05, 0.95, 0.05, 500) ]

# this function defines an experiment (input distribution and datacenter)
# when creating the master experiment, xValues should be zero
# when creating slave experiments, xValues should be extracted from warmed-up master experiment
def createExperiment(xValues = []):

	global stat_config
	
	# setup experiment skeleton
	experimentInput = sqs_core.ExperimentInput()
	experimentOutput = sqs_core.ExperimentOutput()
	
	# insert stat_config into experiment output
	for i in range(0, len(stat_config)):
		if (xValues == []):
			experimentOutput.addOutput(stat_config[i][0], stat_config[i][1], stat_config[i][2], stat_config[i][3], stat_config[i][4])
		else: 
			experimentOutput.addOutput(stat_config[i][0], stat_config[i][1], stat_config[i][2], stat_config[i][3], stat_config[i][4], xValues[i])
	
	rand = sqs_generator.MTRandom(long(1))
	experiment = sqs_core.Experiment("Power capping test", rand, experimentInput, experimentOutput)
	
	# service file used by generators
	arrivalFile = "workloads/www.arrival.cdf"
	serviceFile = "workloads/www.service.cdf"

	# specify input distribution
	cores = 4
	sockets = 1
	targetRho = 0.5

	arrivalDistribution = sqs_math.Distribution.loadDistribution(arrivalFile, 1e-3)
	serviceDistribution = sqs_math.Distribution.loadDistribution(serviceFile, 1e-3)

	averageInterarrival = arrivalDistribution.getMean()
	averageServiceTime = serviceDistribution.getMean()
	qps = 1/averageInterarrival
	rho = qps/(cores*(1/averageServiceTime))
	arrivalScale = rho/targetRho
	averageInterarrival = averageInterarrival*arrivalScale
	serviceRate = 1/averageServiceTime
	scaledQps = (qps/arrivalScale)
	
	# debug output
	#print "Cores: %s" % cores
	#print "rho: %s"	% rho
	#print "recalc rho: %s" % (scaledQps/(cores*(1/averageServiceTime)))
	#print "arrivalScale: %s" % arrivalScale
	#print "Average interarrival time: %s" % averageInterarrival
	#print "QPS as is %s" % qps
	#print "Scaled QPS: %s" % scaledQps
	#print "Service rate as is %s" % serviceRate
	#print "Service rate x: %s" % cores + " is: %s" % ((serviceRate)*cores)
	#print "\n------------------\n"
	
	arrivalGenerator = sqs_generator.EmpiricalGenerator(rand, arrivalDistribution, "arrival", arrivalScale)
	serviceGenerator = sqs_generator.EmpiricalGenerator(rand, serviceDistribution, "service", 1.0)

	#setup datacenter	
	dataCenter = sqs_datacenter.DataCenter()

	nServers = 100	
	capPeriod = 1.0
	globalCap = 70.0 * nServers
	maxPower = 100.0 * nServers
	minPower = 59.0 * nServers
	enforcer = sqs_datacenter.PowerCappingEnforcer(experiment, capPeriod, globalCap, maxPower, minPower)
	for i in range(0, nServers):
		server = sqs_datacenter.Server(sockets, cores, experiment, arrivalGenerator, serviceGenerator)

		server.setSocketPolicy(sqs_SocketPowerPolicy.NO_MANAGEMENT)
		server.setCorePolicy(sqs_CorePowerPolicy.NO_MANAGEMENT)	
		coreActivePower = 40 * (4.0/5)/cores
		coreHaltPower = coreActivePower*0.2
		coreParkPower = 0.0

		socketActivePower = 40 * (1.0/5)/sockets
		socketParkPower = 0.0

		server.setCoreActivePower(coreActivePower)
		server.setCoreParkPower(coreParkPower)
		server.setCoreHaltPower(coreHaltPower)

		server.setSocketActivePower(socketActivePower)
		server.setSocketParkPower(socketParkPower)
		enforcer.addServer(server)
		dataCenter.addServer(server)
		
	experimentInput.addDataCenter(dataCenter)
	
	return experiment

# ========== Don't change code below ==========

# this function runs a distributed experiment
# it passes experiments and results between Python and Java
# don't change this function
def runExperiment(cfg):
	try:
		# instantiate master
		master = sqs_master.Master()

		# load machine config
		numberOfSlaves = master.parseConfigFile(cfg)
		print "Number of Slaves: %s" % numberOfSlaves

		# create master experiment
		masterExperiment = createExperiment()

		# run master experiment to steady state
		master.runMasterExperiment(masterExperiment)

		# extract xValues, distribute and run slave experiments
		xValues = []
		for i in range(0, len(stat_config)):
			xValues.append(masterExperiment.getStats().getStat(stat_config[i][0]).getHistogramXValues());
			
		slaveExperiments = []
		for i in range(0, numberOfSlaves):
			slaveExperiments.append(createExperiment(xValues))
			
		master.runSlaveExperiment(slaveExperiments)
		
		# simulation finish
		time.sleep(10)

	except jpype.JavaException as ex:
		print "Caught Java Exception, exit. "
		print ex.stacktrace()
		exit(1)

def usage():
	print "Usage: ./master.py <machine config>"
	exit(1)
	
def exit(x):
	print "========== JPype Output Begin =========="
	jpype.shutdownJVM()
	print "========== JPype Output End ============"
	sys.exit(x)

def main(argv):
	if len(argv) == 1:
		runExperiment(argv[0])
	else:
		usage()
	exit(0)
	
if __name__ == "__main__":
    main(sys.argv[1:])
